src/json_modbus.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.2.
.TH JSON_MODBUS "1" "April 2016" "json_modbus 1.0.1" "User Commands"
.SH NAME
json_modbus \- JSON Modbus
.SH SYNOPSIS
.B json_modbus
[\fI\,OPTIONS\/\fR]... [\fI\,HOSTNAME\/\fR]...
.SH DESCRIPTION
json_modbus 1.0.1
.PP
v1.0.0_ecowatch\-5\-gc23db4299b\-dirty
.RS
Registers of the modbus device can be passed as options or defined in a configuration file and read in using the -C commandline option

The register string as the following format for reading:
<reg_type> <reg_name> <reg_address>

The register string as the following format for writing:
<reg_type> <reg_name> <reg_address> <reg_value> [conversion]

The conversion parameter is optional

Register types can be one of the following:
.IP \[bu] 2
uint8
.IP \[bu]
uint16
.IP \[bu]
uint32
.IP \[bu]
int8
.IP \[bu]
int16
.IP \[bu]
int32
.IP \[bu]
float
.IP \[bu]
float_cdba
.IP \[bu]
float_array
.RE
.IP \[bu]
coil
.RE
.IP \[bu]
coils (write only!)
.RE

The coils type enables you to write multiple coils at once so it has a slightly differen syntax:

coils <coils_name> <coil_1_val> <coil_2_val> <coil_n_val>
.PP
This application reads a device using the modbus protocol and returns the
result in json format.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.TP
\fB\-p\fR, \fB\-\-port\fR=\fI\,INT\/\fR
Port of modbus device.  (default=`502')
.TP
\fB\-i\fR, \fB\-\-interval\fR=\fI\,INT\/\fR
Time between measurements in seconds  (default=`0')
.TP
\fB\-d\fR, \fB\-\-debug\fR
Show protocol debug information  (default=off)
.TP
\fB\-n\fR, \fB\-\-name\fR=\fI\,STRING\/\fR
Name of the application  (default=`modbus')
.TP
\fB\-\-include\-date\fR
add a date to the output  (default=off)
.TP
\fB\-C\fR, \fB\-\-conf\-file\fR=\fI\,STRING\/\fR
Configuration file
.TP
\fB\-g\fR, \fB\-\-reg\fR=\fI\,STRING\/\fR
Define a register to read or write
.RS
Registers of the modbus device can be passed as options or defined in a configuration file and read in using the -C commandline option

The register string as the following format for reading:
<reg_type> <reg_name> <reg_address>

The register string as the following format for writing:
<reg_type> <reg_name> <reg_address> <reg_value> [conversion]

The conversion parameter is optional

Register types can be one of the following:
.IP \[bu] 2
uint8
.IP \[bu]
uint16
.IP \[bu]
uint32
.IP \[bu]
int8
.IP \[bu]
int16
.IP \[bu]
int32
.IP \[bu]
float
.IP \[bu]
float_cdba
.IP \[bu]
float_array
.RE
.IP \[bu]
coil
.RE
.IP \[bu]
coils (write only!)
.RE

The coils type enables you to write multiple coils at once so it has a slightly differen syntax:

coils <coils_name> <coil_1_val> <coil_2_val> <coil_n_val>
.TP
\fB\-t\fR, \fB\-\-timeout\fR=\fI\,INT\/\fR
Set the response timeout  (default=`1')
.TP
\fB\-s\fR, \fB\-\-serial\fR
Use serial rtu  (default=off)
.TP
\fB\-b\fR, \fB\-\-baud\fR=\fI\,INT\/\fR
Serial baudrate  (default=`115200')
.TP
\fB\-I\fR, \fB\-\-input\fR
Read from input registers not holding  (default=off)
.TP
\fB\-S\fR, \fB\-\-swap\fR
Swap two 16 bits in (u)int32  (default=off)
.TP
\fB\-\-delay\fR=\fI\,INT\/\fR
Delay before starting to send  (default=`0')
.IP
Mode: read
.TP
\fB\-r\fR, \fB\-\-read\fR
Read registers  (default=on)
.IP
Mode: write
.TP
\fB\-w\fR, \fB\-\-write\fR
Write registers  (default=off)
.PP
v1.0.0_ecowatch\-5\-gc23db4299b\-dirty
.RS
Registers of the modbus device can be passed as options or defined in a configuration file and read in using the -C commandline option

The register string as the following format for reading:
<reg_type> <reg_name> <reg_address>

The register string as the following format for writing:
<reg_type> <reg_name> <reg_address> <reg_value> [conversion]

The conversion parameter is optional

Register types can be one of the following:
.IP \[bu] 2
uint8
.IP \[bu]
uint16
.IP \[bu]
uint32
.IP \[bu]
int8
.IP \[bu]
int16
.IP \[bu]
int32
.IP \[bu]
float
.IP \[bu]
float_cdba
.IP \[bu]
float_array
.RE
.IP \[bu]
coil
.RE
.IP \[bu]
coils (write only!)
.RE

The coils type enables you to write multiple coils at once so it has a slightly differen syntax:

coils <coils_name> <coil_1_val> <coil_2_val> <coil_n_val>
.SH EXAMPLES
# Read coil on address 1 and give the response label coil_name

json_modbus -g 'coil coil_name 1' -r -h example.com -p 1502

outputs:
.br
{
.br
	"modbus": {
.br
		"coil_name": 0
.br
	}
.br
}

# Writing a uint32 register
json_modbus -g 'uint reg_name 10 1000000' -w -h example.com -p 1502
.SH "SEE ALSO"
The full documentation for
.B json_modbus
is maintained as a Texinfo manual.  If the
.B info
and
.B json_modbus
programs are properly installed at your site, the command
.IP
.B info json_modbus
.PP
should give you access to the complete manual.


src/json_modbus.c
#include <modbus.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdbool.h>
#include <errno.h>
#include <time.h>
#include "cmdline.h"
#include <signal.h>
#include <regex.h>
#include "dbg.h"
#include "json_modbus.h"

//#define address_list_size (int)(sizeof(address_list)/sizeof(address_list[0]))


static void catch_function(int signo) {
        if (exit_now) {
                DEBUG_MSG("Exiting now!\n");
                //modbus_close(ctx);
                //modbus_free(ctx);
                exit(-1);
        } else {
                if (signo == SIGINT) {
                        DEBUG_MSG("Exiting nicely\n");
                        running = false;
                        exit_now = true;
                }
        }

}

void parse_def_string(char * def_str,struct mb_util_ctx * ctx) {
        char type[16];
        char name[51];
        uint16_t reg;
        char value[51];
        float conversion = 1;
        int last_pos = 0;
        ctx->reg_list[ctx->reg_index].convert = false;
        // --reg "type name reg [value] conversion"

#define STRLEN(s) ((sizeof(s)/sizeof(s[0]))-1)
#define CHECK_STR(str,sstr) (strncmp(str,sstr,STRLEN(sstr)) == 0)

        if (CHECK_STR(def_str,"int") || CHECK_STR(def_str,"uint") || CHECK_STR(def_str,"float") || CHECK_STR(def_str,"float_cdab")) {
                if (ctx->rw == 'r') {
                        sscanf(def_str, "%15s%50s%hu%n",type,name,&reg,&last_pos);
                        DEBUG_MSG("Lst pos: %d size: %lu\n",last_pos, strlen(def_str));
                } else {
                        sscanf(def_str, "%15s%50s%hu%20s%n",type,name,&reg,value,&last_pos);
                }

                if (last_pos != strlen(def_str)) {
                        sscanf(&def_str[last_pos], "%f",&conversion);
                        ctx->reg_list[ctx->reg_index].convert = true;
                }
        }

        if (CHECK_STR(def_str,"int8")) {
                ctx->reg_list[ctx->reg_index].type = mb_int8;
                ctx->reg_list[ctx->reg_index].int8_val = (int8_t)atoi(value);
        } else if (CHECK_STR(def_str,"int16")) {
                ctx->reg_list[ctx->reg_index].type = mb_int16;
                ctx->reg_list[ctx->reg_index].int16_val = (int16_t)atoi(value);
        } else if (CHECK_STR(def_str,"int32")) {
                ctx->reg_list[ctx->reg_index].type = mb_int32;
                ctx->reg_list[ctx->reg_index].uint32_val = (int32_t)atoi(value);
        } else if (CHECK_STR(def_str,"uint8")) {
                ctx->reg_list[ctx->reg_index].type = mb_uint8;
                ctx->reg_list[ctx->reg_index].uint8_val = (uint8_t)atoi(value);
        } else if (CHECK_STR(def_str,"uint16")) {
                ctx->reg_list[ctx->reg_index].type = mb_uint16;
                ctx->reg_list[ctx->reg_index].uint16_val = (uint16_t)atoi(value);
        } else if (CHECK_STR(def_str,"uint32")) {
                ctx->reg_list[ctx->reg_index].type = mb_uint32;
                ctx->reg_list[ctx->reg_index].uint32_val = (uint32_t)atoi(value);
        } else if (CHECK_STR(def_str,"float")) {
                ctx->reg_list[ctx->reg_index].type = mb_float;
                ctx->reg_list[ctx->reg_index].float_val = atof(value);
        } else if (CHECK_STR(def_str,"float_cdab")) {
                ctx->reg_list[ctx->reg_index].type = mb_float_cdab;
                ctx->reg_list[ctx->reg_index].float_val = atof(value);
        } else if (CHECK_STR(def_str,"float_array")) {
                ctx->reg_list[ctx->reg_index].type = mb_float_array;
        } else if (CHECK_STR(def_str,"coils")) {
                sscanf(def_str, "%15s%50s%hu",type,name,&reg);
                ctx->reg_list[ctx->reg_index].type = mb_coils;
                char *tok = NULL;
                tok = strtok(&def_str[5], " ");
                int i;
                DEBUG_MSG("str %s\n",def_str);
                while (tok) {
                        DEBUG_MSG("Token: %s\n", tok);
                        ctx->reg_list[ctx->reg_index].coil_array[i++] = (bool)atoi(tok);
                        tok = strtok(NULL, " ");
                }
                DEBUG_MSG("NUM coils = %d\n",i);
                ctx->reg_list[ctx->reg_index].num_coils = i;

        } else if (CHECK_STR(def_str,"coil")) {
                sscanf(def_str, "%15s%50s%hu%10s",type,name,&reg,value);
                ctx->reg_list[ctx->reg_index].type = mb_coil;
                ctx->reg_list[ctx->reg_index].uint8_val = (uint8_t)atoi(value);
        }

        ctx->reg_list[ctx->reg_index].address = reg;
        strncpy(ctx->reg_list[ctx->reg_index].name,name,50);
        ctx->reg_list[ctx->reg_index].conversion = conversion;

        ctx->reg_index++;
        DEBUG_MSG("Type: %s\nName: %s\nReg: %u\nValue: %s\nConversion: %f\n",type,name,reg,value,conversion);

}

void process_registers(struct mb_util_ctx * ctx) {
        uint16_t tmp[2];
        DEBUG_MSG("Process registers: %d\n",ctx->reg_index);
        DEBUG_MSG("Reg list rw: %c\n",ctx->rw);

        // Read registers
        if (ctx->rw == 'r') {
                for (int i=0; i<ctx->reg_index; i++) {
                    usleep(5000);
                        DEBUG_MSG("Reading: reg: %hu\n",ctx->reg_list[i].address);
                        switch(ctx->reg_list[i].type) {
                        case mb_int8:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 1, (uint16_t *)&ctx->reg_list[i].int8_val);
                                break;
                        case mb_int16:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 1, (uint16_t *)&ctx->reg_list[i].int16_val);
                                break;
                        case mb_int32:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 2, (uint16_t *)&ctx->reg_list[i].int32_val);
                                break;
                        case mb_uint8:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 1, (uint16_t *)&ctx->reg_list[i].int8_val);
                                break;
                        case mb_uint16:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 1, &ctx->reg_list[i].uint16_val);
                                break;
                        case mb_uint32:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 2, (uint16_t *)&ctx->reg_list[i].uint32_val);
                                break;
                        case mb_float:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 2, tmp);
                                ctx->reg_list[i].float_val = modbus_get_float(tmp);
                                break;
                        case mb_float_cdab:
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 2, tmp);
                                ctx->reg_list[i].float_val = modbus_get_float_cdab(tmp);
                                break;
                        case mb_float_array:
                                // Allocate room for float array
                                ctx->reg_list[i].float_array = (uint16_t *)malloc(sizeof(uint16_t) * ctx->reg_list[i].conversion);
                                rc = ctx->read_regs(ctx->modbus_ctx, ctx->reg_list[i].address, 2 * ctx->reg_list[i].conversion, ctx->reg_list[i].float_array);
                                //ctx->reg_list[i].float_val = modbus_get_float_cdab(tmp);
                                break;
                        case mb_coil:
                                rc = modbus_read_bits(ctx->modbus_ctx, ctx->reg_list[i].address, 1, &ctx->reg_list[i].uint8_val);
                                break;
                        default:
                                break;
                        }
                }
        } else { // Write registers
                for (int i=0; i<ctx->reg_index; i++) {
                    usleep(5000);
                        uint16_t tmp_float[2];
                        DEBUG_MSG("Writing: reg: %hu\n",ctx->reg_list[i].address);
                        switch(ctx->reg_list[i].type) {
                        case mb_uint16:
                                rc = modbus_write_register(ctx->modbus_ctx, ctx->reg_list[i].address, ctx->reg_list[i].uint16_val);
                                break;
                        case mb_uint32:
                                rc = modbus_write_registers(ctx->modbus_ctx, ctx->reg_list[i].address, 2,(uint16_t*)&ctx->reg_list[i].uint32_val);
                                break;
                        case mb_int16:
                                rc = modbus_write_register(ctx->modbus_ctx, ctx->reg_list[i].address, (uint16_t)ctx->reg_list[i].int16_val);
                                break;
                        case mb_int8:
                                rc = modbus_write_register(ctx->modbus_ctx, ctx->reg_list[i].address, (uint16_t)ctx->reg_list[i].int8_val);
                                break;
                        case mb_uint8:
                                rc = modbus_write_register(ctx->modbus_ctx, ctx->reg_list[i].address, (uint16_t)ctx->reg_list[i].int8_val);
                                break;
                        case mb_float:
                                modbus_set_float(ctx->reg_list[i].float_val,tmp_float);
                                rc = modbus_write_registers(ctx->modbus_ctx, ctx->reg_list[i].address, 2, tmp_float);
                                break;
                        case mb_float_cdab:
                        case mb_float_array:
                                DEBUG_MSG("writing float not implemented");
                                exit(1);
                                break;
                        case mb_coil:
                                rc = modbus_write_bit(ctx->modbus_ctx, ctx->reg_list[i].address, ctx->reg_list[i].uint8_val);
                                break;
                        case mb_coils:
                                rc = modbus_write_bits(ctx->modbus_ctx, ctx->reg_list[i].address, ctx->reg_list[i].num_coils,(uint8_t *)ctx->reg_list[i].coil_array);
                                break;
                        default:
                                break;
                        }
                }
        }
        if (rc == -1) {
            printf("{\n");
            printf("\t\"error\": \"%s\"\n",modbus_strerror(errno));
            printf("}\n");
            exit(1);
        }
}

void print_registers(struct mb_util_ctx * ctx) {
        //printf("Process registers: %d\n",ctx->reg_index);
        printf("{\n");
        printf("\t\"%s\": {\n", ctx->name);
        if(ctx->include_date){
            char buffer[26];
            time_t mytime = time(NULL);
            struct tm* tm_info;
            tm_info = localtime(&mytime);
            strftime(buffer, 26, "%Y:%m:%d %H:%M:%S", tm_info);
            printf("\t\t\"TimeInfo\":");
            printf("\"%s\",\n",buffer);
        }

        for (int i=0; i<ctx->reg_index; i++) {
                //printf("Reg list rw: %c\n",ctx->reg_list[i].rw);
                //printf("Reading: reg: %hu\n",ctx->reg_list[i].address);

                if (ctx->reg_list[i].convert && ctx->reg_list[i].type != mb_float_array) {
                        ctx->reg_list[i].float_val *= ctx->reg_list[i].conversion;
                        printf("\t\t\"%s\":%f", ctx->reg_list[i].name, ctx->reg_list[i].float_val);
                } else {
                        switch(ctx->reg_list[i].type) {
                        case mb_int8:
                                printf("\t\t\"%s\":%hhd", ctx->reg_list[i].name, ctx->reg_list[i].int8_val);
                                break;
                        case mb_int16:
                                //if (ctx->reg_list[i].convert) ctx->reg_list[i].int16_val *= ctx->reg_list[i].conversion;
                                printf("\t\t\"%s\":%hd", ctx->reg_list[i].name, ctx->reg_list[i].int16_val);
                                break;
                        case mb_int32:
                                //if (ctx->reg_list[i].convert) ctx->reg_list[i].int32_val *= ctx->reg_list[i].conversion;
                                if(ctx->swap){
                                ctx->reg_list[i].int32_val=(((ctx->reg_list[i].int32_val & 0x0000FFFF)<<16)|((ctx->reg_list[i].int32_val & 0xFFFF0000))>>16);
                                }
                                printf("\t\t\"%s\":%d", ctx->reg_list[i].name, ctx->reg_list[i].int32_val);
                                break;
                        case mb_uint8:
                                //if (ctx->reg_list[i].convert) ctx->reg_list[i].uint8_val *= ctx->reg_list[i].conversion;
                                printf("\t\t\"%s\":%hhu", ctx->reg_list[i].name, ctx->reg_list[i].uint8_val );
                                break;
                        case mb_uint16:
                                //if (ctx->reg_list[i].convert) ctx->reg_list[i].uint16_val *= ctx->reg_list[i].conversion;
                                printf("\t\t\"%s\":%hu", ctx->reg_list[i].name, ctx->reg_list[i].uint16_val);
                                break;
                        case mb_uint32:
                                //if (ctx->reg_list[i].convert) ctx->reg_list[i].uint32_val *= ctx->reg_list[i].conversion;
                                if(ctx->swap){
                                ctx->reg_list[i].uint32_val=(((ctx->reg_list[i].uint32_val & 0x0000FFFF)<<16)|((ctx->reg_list[i].uint32_val & 0xFFFF0000))>>16);
                                }
                                printf("\t\t\"%s\":%u", ctx->reg_list[i].name, ctx->reg_list[i].uint32_val);
                                break;
                        case mb_float:
                        case mb_float_cdab:
                                //if (ctx->reg_list[i].convert) ctx->reg_list[i].float_val *= ctx->reg_list[i].conversion;
                                printf("\t\t\"%s\":%f", ctx->reg_list[i].name, ctx->reg_list[i].float_val);
                                break;
                        case mb_float_array:
                                printf("\t\t\"%s\":[", ctx->reg_list[i].name);
                                for (int j=0; j<ctx->reg_list[i].conversion; j++) {
                                        printf("%f",modbus_get_float(ctx->reg_list[i].float_array+(j*2)));
                                        if (j != ctx->reg_list[i].conversion - 1) {
                                                printf(",");
                                        }
                                }
                                printf("]");
                                // Cleanup memory
                                free(ctx->reg_list[i].float_array);
                                break;
                        case mb_coil:
                                printf("\t\t\"%s\":%hhu", ctx->reg_list[i].name, ctx->reg_list[i].uint8_val);
                                break;
                        case mb_coils:
                                printf("\t\t\"%s\":[", ctx->reg_list[i].name);
                                for (int j=0; j<ctx->reg_list[i].num_coils; j++) {
                                        printf("%hhu",(uint8_t)ctx->reg_list[i].coil_array[j]);
                                        if (j != ctx->reg_list[i].num_coils - 1) {
                                                printf(",");
                                        }
                                }
                                printf("]");
                                break;
                        default:
                                break;
                        }
                }

                // Last item in list
                if (i != ctx->reg_index - 1) {
                        printf(",\n");
                } else {
                        printf("\n");
                }
        }
     printf("\t}\n");
     printf("}\n");
}

int main(int argc, char **argv) {
        if (signal(SIGINT, catch_function) == SIG_ERR) {
                DEBUG_MSG("An error occurred while setting a signal handler.\n");
                return EXIT_FAILURE;
        }

        struct cmdline_parser_params *params;

        /* initialize the parameters structure */
        params = cmdline_parser_params_create();

        params->initialize = 1;
        params->override = 1;
        params->check_required = 0;

        /* let's call our cmdline parser */
        if (cmdline_parser_ext (argc, argv, &args_info, params) != 0) exit(1);

        params->initialize = 0;
        params->override = 0;
        params->check_required = 1;

        if (args_info.conf_file_given) {
                DEBUG_MSG("Reading config\n");
                /* call the config file parser */
                if (cmdline_parser_config_file(args_info.conf_file_arg, &args_info, params) != 0) exit(1);
        }

        // if (args_info.include_date_given) {
        //                       /* call the config file parser */
        //         if (cmdline_parser_config_file(args_info.conf_file_arg, &args_info, params) != 0) exit(1);
        // }

        if (args_info.inputs_num != 1) {
                cmdline_parser_print_help();
                exit(1);
        }
        debug_mode = args_info.debug_flag;

        struct mb_util_ctx mb_instance;
        mb_instance.reg_list = (reg_list_t *) malloc(sizeof(reg_list_t) * args_info.reg_given);
        mb_instance.reg_index = 0;
        mb_instance.swap = args_info.swap_given;
        mb_instance.include_date = args_info.include_date_given;

        if (args_info.write_flag == true) {
                DEBUG_MSG("Write registers\n");
                mb_instance.rw = 'w';
        } else {
                DEBUG_MSG("Read registers\n");
                mb_instance.rw = 'r';
        }
        // Copy system name
        strncpy(mb_instance.name,args_info.name_arg,49);


        for (int i = 0; i < args_info.reg_given; ++i) {
                parse_def_string(args_info.reg_arg[i],&mb_instance);
                DEBUG_MSG("passed coil: %s\n", args_info.reg_arg[i]);
        }

        if(args_info.serial_given) {
                mb_instance.modbus_ctx = modbus_new_rtu(args_info.inputs[0],args_info.baud_arg,'N',8,2);
        } else {
                DEBUG_MSG("Connecting to: %s\n",args_info.inputs[0]);
                mb_instance.modbus_ctx = modbus_new_tcp(args_info.inputs[0], args_info.port_arg);
        }

        if (mb_instance.modbus_ctx == NULL) {
                DEBUG_MSG("Unable to create the libmodbus context\n");
                return -1;
        }
        modbus_set_slave(mb_instance.modbus_ctx, 1);
        modbus_set_debug(mb_instance.modbus_ctx,args_info.debug_flag);

        if (args_info.input_given) {
          mb_instance.read_regs = modbus_read_input_registers;
        } else {
          mb_instance.read_regs = modbus_read_registers;
        }

        struct timeval rt;

        /*rt2.tv_sec=1;
           rt2.tv_usec=0;
           modbus_set_byte_timeout(mb_instance.modbus_ctx, &rt);
           rt.tv_sec=10;
           rt.tv_usec=0;
           modbus_set_response_timeout(mb_instance.modbus_ctx, &rt);
         */
        rt.tv_sec=args_info.timeout_arg;
        rt.tv_usec=0;
        modbus_set_response_timeout(mb_instance.modbus_ctx, &rt);
        modbus_set_byte_timeout(mb_instance.modbus_ctx, &rt);

        if (modbus_connect(mb_instance.modbus_ctx) == -1) {
                DEBUG_MSG("Connection failed: %s\n", modbus_strerror(errno));
                modbus_free(mb_instance.modbus_ctx);
                return -1;
        }
        sleep(args_info.delay_arg);

        process_registers(&mb_instance);

        print_registers(&mb_instance);

        modbus_close(mb_instance.modbus_ctx);
        modbus_free(mb_instance.modbus_ctx);
        free(mb_instance.reg_list);

        return 0;
}

/* Get a float from 4 bytes in Modbus format (ABCD) */
float modbus_get_float_cdab(const uint16_t *src)
{
        float f;
        uint32_t i;

        i = (((uint32_t)src[0]) << 16) + src[1];
        memcpy(&f, &i, sizeof(float));

        return f;
}
